src/wire_msg.rs

```diff
@@ -22,25 +20,27 @@ pub enum WireMsg {
     UserMsg(bytes::Bytes),
 }
 
-impl Into<bytes::Bytes> for WireMsg {
-    fn into(self) -> bytes::Bytes {
-        if let WireMsg::UserMsg(ref m) = self {
-            if m.len() > MAX_MESSAGE_SIZE_FOR_SERIALISATION {
-                return m.clone();
-            }
-        }
+const USER_MESSAGE: u8 = 0;
+const WIRE_MESSAGE: u8 = 1;
 
-        From::from(unwrap!(bincode::serialize(&self)))
+impl Into<(u8, bytes::Bytes)> for WireMsg {
+    fn into(self) -> (u8, bytes::Bytes) {
+        match self {
+            WireMsg::UserMsg(ref m) => (USER_MESSAGE, m.clone()),
```

Comment:
`From` and `Into` are asymmetric. Maybe better to return `(m.clone, USER_MSG_FLAG)` since the flag is the last byte. For the other branch it will ofc be `(serialize(...), !USER_MSG_FLAG)`.